GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,216 @@
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.Tracing
+{
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private const int DEFAULT_MAX_QUEUE_SIZE = 256;
+
+        private readonly Func<NamedPipeClientStream> createPipeFunc;
+        private readonly BlockingCollection<string> queue;
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+
+        // An integer is used here rather than the enum so that it can be used with Interlocked.Exchange(..)
+        private int state = (int)QueuedPipeStringWriterState.Unknown;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> createPipeFunc, int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE)
+        {
+            this.createPipeFunc = createPipeFunc;
+            this.queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public event EventHandler<QueuedPipeStringWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeStringWriterState
+        {
+            Unknown = 0,
+            Stopped = 1,
+            Failing = 2,
+            Healthy = 3,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeStringWriter)} is not started");
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.isRunning = false;
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
```

Comment:
Why is it important to drain whatever is one the queue?  Would we refuse new requests?

Reply:
It would be odd for the following to be true:

```csharp
using (var writer = new QueuedPipeStringWriter(..))
{
    writer.Start();
    writer.Enqueue("message1");
    writer.Enqueue("message2");
    writer.Stop();
}

// Pipe gets message1, but not message2 because it had not been sent before Stop().
```

Reply:
That's fair enough, I have some questions on handling when we can't add.